server/base/src/main/java/org/apache/accumulo/server/metadata/iterators/TabletMetadataCheckIterator.java

```diff
@@ -0,0 +1,191 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.accumulo.server.metadata.iterators;
+
+import static org.apache.accumulo.core.util.LazySingletons.GSON;
+import static org.apache.accumulo.server.metadata.iterators.ColumnFamilyTransformationIterator.getTabletRow;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.EnumSet;
+import java.util.Map;
+import java.util.NoSuchElementException;
+import java.util.Objects;
+import java.util.Set;
+
+import org.apache.accumulo.core.classloader.ClassLoaderUtil;
+import org.apache.accumulo.core.client.IteratorSetting;
+import org.apache.accumulo.core.data.ByteSequence;
+import org.apache.accumulo.core.data.Condition;
+import org.apache.accumulo.core.data.Key;
+import org.apache.accumulo.core.data.Range;
+import org.apache.accumulo.core.data.Value;
+import org.apache.accumulo.core.dataImpl.KeyExtent;
+import org.apache.accumulo.core.iterators.IteratorAdapter;
+import org.apache.accumulo.core.iterators.IteratorEnvironment;
+import org.apache.accumulo.core.iterators.SortedKeyValueIterator;
+import org.apache.accumulo.core.metadata.schema.TabletMetadata;
+import org.apache.accumulo.core.metadata.schema.TabletMetadataCheck;
+import org.apache.accumulo.server.metadata.ConditionalTabletMutatorImpl;
+import org.apache.hadoop.io.Text;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.base.Preconditions;
+
+public class TabletMetadataCheckIterator implements SortedKeyValueIterator<Key,Value> {
+
+  private static final Logger log = LoggerFactory.getLogger(TabletMetadataCheckIterator.class);
+
+  private TabletMetadataCheck check;
+  private KeyExtent expectedExtent;
+
+  private Key startKey;
+  private Value topValue;
+  private SortedKeyValueIterator<Key,Value> source;
+
+  private static final String CHECK_CLASS_KEY = "checkClass";
+  private static final String CHECK_DATA_KEY = "checkData";
+  private static final String EXTENT_KEY = "checkExtent";
+  private static final String SUCCESS = "success";
+
+  @Override
+  public void init(SortedKeyValueIterator<Key,Value> source, Map<String,String> options,
+      IteratorEnvironment env) throws IOException {
+    Preconditions.checkState(check == null && expectedExtent == null && startKey == null);
+    try {
+      String className = Objects.requireNonNull(options.get(CHECK_CLASS_KEY));
+      String checkData = Objects.requireNonNull(options.get(CHECK_DATA_KEY));
+      log.trace("Instantiating class {} using {}", className, checkData);
+      Class<? extends TabletMetadataCheck> clazz =
+          ClassLoaderUtil.loadClass(null, className, TabletMetadataCheck.class);
+      check = GSON.get().fromJson(checkData, clazz);
+      expectedExtent = KeyExtent.fromBase64(options.get(EXTENT_KEY));
+      this.source = source;
+    } catch (ReflectiveOperationException e) {
+      throw new IllegalStateException(e);
```

Comment:
Would it be useful to add any logging here for a ClassNotFoundException if the impl isn't on the server classpath? It may not be necessary if this exception will be logged higher up when the IllegalStateException is thrown.

Reply:
It depends on if the existing exception messages include the class name.  Could add the class name to the IllegalStateException to be sure.